### ISSUE TITLE
Pattern matcher fails if query is in another atomspace than kb

### ISSUE THREAD

**Issue by ngeiswei** [opencog/atomspace#1387]
The following script highlights the problem
```
(use-modules (opencog query))

;; Define query in the initial atomspace
(define query
(BindLink
  (EvaluationLink
    (PredicateNode "P")
    (VariableNode "$X"))
  (Concept "A"))
)

;; Switch to kb-as
(define kb-as (cog-new-atomspace))
(cog-set-atomspace! kb-as) ;; <--- bug

(EvaluationLink (stv 1 1)
   (PredicateNode "P")
   (Concept "my-arg"))             

(cog-bind query)
```
It returns the empty Set, only returns the solution `(Concept "A")` if the current atomspace isn't set to `kb-as`.

**Comment by ngeiswei** [comment]
@linas I think I know what the problem is, `InitiateSearchCB::neighbor_search` considers the incoming set of `(PredicateNode "P")` in the atomspace of the query instead of `kb-as`. But just to be sure, it's a bug, not a feature, right?

**Comment by linas** [comment]
Its a feature, more-or-less. By declaring `kb-as` to be a new atomspace, instead of a derived one, you've created two distinct atoms that are `(PredicateNode "P")`  These two atoms have two distinct incoming sets, and so the exploration of the connected graph doesn't find them to be attached.

Think of it like this: the atomspace is an implicit part of the name of the atom, so you really have  `(PredicateNode "P in primary-atomspace")`   and  `(PredicateNode "P in other atomspace")`  and now its clear that these really are two different atoms. So of course the patten matcher rejects the second "P" -- its clearly a mismatch.

The above should make clear that this has nothing to do with the pattern matcher - its really about how multiple atomspaces are defined. The pattern miner, and any other block of code will all trip over this same issue.

**Comment by ngeiswei** [comment]
I'm still feeling a bit ambivalent about it, I suppose time will tell if it is a bug or a feature.

However, in the immediate term we want to add some explanation somewhere (if it's not already present), log a warning, etc, right?

**Comment by ngeiswei** [comment]
Suppose atom `C` is in 2 disconnected atomspaces `A1` and `A2`, let `h1` and `h2` be the handles of `C` in `A1` and `A2` respectively. Observe that `A2.get_atom(h1)` returns `h2`.

To be consistent with that, I think it makes sense to allow mining atomspace `A2` with a query from atomspace `A1`.

**Comment by ngeiswei** [comment]
I'm closing, if one really wants to have the query not pollute the KB it can put it in a children atomspace thereof.

**Comment by ngeiswei** [comment]
Let me re-open, I believe a user warning should be logged when attempting to do that, otherwise it will puzzle more than one user.

**Comment by linas** [comment]
Closing, this is fixed in #2661
